README.md
# PayPal Payouts API SDK for NodeJS

![PayPal Developer](homepage.jpg)

__Welcome to PayPal NodeJS SDK__. This repository contains PayPal's NodeJS SDK and samples for [v1/payments/payouts](https://developer.paypal.com/docs/api/payments.payouts-batch/v1/) APIs.

This is a part of the next major PayPal SDK. It includes a simplified interface to only provide simple model objects and blueprints for HTTP calls. This repo currently contains functionality for PayPal Payouts APIs which includes [Payouts](https://developer.paypal.com/docs/api/payments.payouts-batch/v1/).

Please refer to the [PayPal Payouts Integration Guide](https://developer.paypal.com/docs/payouts/) for more information. Also refer to [Setup your SDK](https://developer.paypal.com/docs/payouts/reference/setup-sdk) for additional information about setting up the SDK's.

## Usage
### Binaries

It is not mandatory to fork this repository for using the PayPal SDK. You can refer [PayPal Payouts SDK](https://developer.paypal.com/docs/payouts/reference/setup-sdk/#install-the-sdk) for configuring and working with SDK without forking this code.

For contributing or referring the samples, You can fork/refer this repository. 

## Examples
### Creating a Payouts
This will create a Payout and print the batch_id for the Payout.
#### Code to Execute:
```javascript
const paypal = require('@paypal/payouts-sdk');
  
// Creating an environment
let clientId = "<<PAYPAL-CLIENT-ID>>";
let clientSecret = "<<PAYPAL-CLIENT-SECRET>>";
let environment = new paypal.core.SandboxEnvironment(clientId, clientSecret);
let client = new paypal.core.PayPalHttpClient(environment);

let requestBody = {
    "sender_batch_header": {
      "recipient_type": "EMAIL",
      "email_message": "SDK payouts test txn",
      "note": "Enjoy your Payout!!",
      "sender_batch_id": "Test_sdk_1",
      "email_subject": "This is a test transaction from SDK"
    },
    "items": [{
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-1@paypal.com",
      "sender_item_id": "Test_txn_1"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-2@paypal.com",
      "sender_item_id": "Test_txn_2"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-3@paypal.com",
      "sender_item_id": "Test_txn_3"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-4@paypal.com",
      "sender_item_id": "Test_txn_4"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-5@paypal.com",
      "sender_item_id": "Test_txn_5"
    }]
  }

// Construct a request object and set desired parameters
// Here, PayoutsPostRequest() creates a POST request to /v1/payments/payouts
let request = new paypal.payouts.PayoutsPostRequest();
request.requestBody(requestBody);

// Call API with your client and get a response for your call
let createPayouts  = async function(){
        let response = await client.execute(request);
        console.log(`Response: ${JSON.stringify(response)}`);
        // If call returns body in response, you can get the deserialized version from the result attribute of the response.
        console.log(`Payouts Create Response: ${JSON.stringify(response.result)}`);
}
createPayouts();
```

### Handle API Failure
This will create a Payout with validation failure to showcase how to parse the failed response entity. Refer samples for more scenarios
#### Code to Execute:
```javascript
const paypal = require('@paypal/payouts-sdk');
  
// Creating an environment
let clientId = "<<PAYPAL-CLIENT-ID>>";
let clientSecret = "<<PAYPAL-CLIENT-SECRET>>";
let environment = new paypal.core.SandboxEnvironment(clientId, clientSecret);
let client = new paypal.core.PayPalHttpClient(environment);

let requestBody = {
    "sender_batch_header": {
      "recipient_type": "EMAIL",
      "email_message": "SDK payouts test txn",
      "note": "Enjoy your Payout!!",
      "sender_batch_id": "Test_sdk_fail",
      "email_subject": "This is a test transaction from SDK"
    },
    "items": [{
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-1@paypal.com",
      "sender_item_id": "Test_txn_1"
    }]
  }

// Construct a request object and set desired parameters
// Here, PayoutsPostRequest() creates a POST request to /v1/payments/payouts
let request = new paypal.payouts.PayoutsPostRequest();
request.requestBody(requestBody);

// Call API with your client and get a response for your call
let createPayouts  = async function(){
    try {
        let response = await client.execute(request);
        console.log(`Response: ${JSON.stringify(response)}`);
        // If call returns body in response, you can get the deserialized version from the result attribute of the response.
        console.log(`Payouts Create Response: ${JSON.stringify(response.result)}`);
    catch {
      if (e.statusCode) {
        console.log("Status code: ", e.statusCode);
        // Parse failure response to get the reason for failure
        const error = JSON.parse(e.message)
        console.log("Failure response: ", error)
        console.log("Headers: ", e.headers)
      } else {
        console.log(e)
      }
    }
}
createPayouts();
```

## Retrieve a Payout Batch
Pass the batchId from the previous sample to retrieve Payouts batch details
### Code to Execute:
```javascript
let getPayouts =  async function(batchId) {
    request = new paypal.payouts.PayoutsGetRequest(batchId);
    request.page(1);
    request.pageSize(10);
    request.totalRequired(true);
    // Call API with your client and get a response for your call
    let response = await client.execute(request);
    console.log(`Response: ${JSON.stringify(response)}`);
    // If call returns body in response, you can get the deserialized version from the result attribute of the response.
    console.log(`Payouts Batch: ${JSON.stringify(response.result)}`);
}

getPayouts('REPLACE-WITH-BATCH-ID'); 
```

## Running tests

To run integration tests using your client id and secret, clone this repository and run the following command:
```sh
$ npm install
$ PAYPAL_CLIENT_ID=YOUR_SANDBOX_CLIENT_ID PAYPAL_CLIENT_SECRET=YOUR_SANDBOX_CLIENT_SECRET npm test
```

## Samples

You can start off by trying out [Payouts Samples](samples/runAll.js)

To run samples run the following command:
```sh
$ npm install
$ PAYPAL_CLIENT_ID=YOUR_SANDBOX_CLIENT_ID PAYPAL_CLIENT_SECRET=YOUR_SANDBOX_CLIENT_SECRET npm test
```

To try out different samples head to [this link](samples)

Note: Update the `payPalClient.js` with your sandbox  credentials or pass your client credentials as environment variable while executing the samples.

## Note

PayPalHttpClient used as part of this project returns Promises

You can read more about Promises here: https://www.promisejs.org/


## License
Code released under [SDK LICENSE](LICENSE)  

CHANGELOG.md
### 1.1.0 (2021-09-21)
*  Updated PayPalHttp package from version 1.0.0 -> 1.0.1

### 1.1.0 (2021-02-19)

*  Specify files to be included in package, removes samples, spec and unwanted files from package ([21527b3](https://github.com/paypal/Payouts-NodeJS-SDK/pull/4/commits/21527b323888037c68fc976c6d988c843f819b9f))
*  Add samples for handling api failure ([8b2f2b7](https://github.com/paypal/Payouts-NodeJS-SDK/commit/8b2f2b70ac60f18bde038a46c11408547d671e30))

spec/payouts/payoutsPostTest.js
'use strict';

require('../spec_helper');
const chai = require('chai');
const client = require('../testHarness').client();
const CreatePayoutsRequest = paypal.payouts.PayoutsPostRequest;

function buildRequestBody() {
  let senderBatchId = "Test_sdk_" + Math.random().toString(36).substring(7);
  return {
    "sender_batch_header": {
      "recipient_type": "EMAIL",
      "email_message": "SDK payouts test txn",
      "note": "Enjoy your Payout!!",
      "sender_batch_id": senderBatchId,
      "email_subject": "This is a test transaction from SDK"
    },
    "items": [{
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-1@paypal.com",
      "sender_item_id": "Test_txn_1"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-2@paypal.com",
      "sender_item_id": "Test_txn_2"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-3@paypal.com",
      "sender_item_id": "Test_txn_3"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-4@paypal.com",
      "sender_item_id": "Test_txn_4"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": "1.00"
      },
      "receiver": "payout-sdk-5@paypal.com",
      "sender_item_id": "Test_txn_5"
    }]
  }
}

function createPayouts() {
  let request = new CreatePayoutsRequest();
  request.requestBody(buildRequestBody());
  return client.execute(request);
}

describe('CreatePayoutsReqeust', function () {
  it('creates a Payouts Batch', function () {
    return createPayouts()
      .then((createResponse) => {
        chai.assert.equal(createResponse.statusCode, 201);
        chai.assert.isNotNull(createResponse.result);

        let responseBody = createResponse.result;
        chai.assert.isNotNull(responseBody.batch_header.payout_batch_id);
        chai.assert.isNotNull(responseBody.batch_header.batch_status);
        chai.assert.equal("This is a test transaction from SDK", responseBody.batch_header.sender_batch_header.email_subject);
        chai.assert.equal("SDK payouts test txn", responseBody.batch_header.sender_batch_header.email_message);

        console.log(JSON.stringify(responseBody, null, 4));
      });
  });
});

module.exports = {
  CreatePayouts: createPayouts
};


samples/runAll.js
const createPayout = require('./createPayout').createPayout;
const createPayoutFailure = require('./createPayout').createPayoutFailure;
const getPayout = require('./getPayout').getPayout;
const getPayoutItem = require('./getPayoutItem').getPayoutItem;
const cancelPayoutItem = require('./cancelPayoutItem').cancelPayoutItem;

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

(async () => {
    //Create Payout
    console.log("Creating Payout...");
    let createResponse = await createPayout(true);
    if (createResponse.statusCode === 201) {
        //Retrieve Payout Batch details
        let payoutId = createResponse.result.batch_header.payout_batch_id;
        console.log("Retrieving Payout details for id - " + payoutId);
        let getResponse = await getPayout(payoutId, true);
        if (getResponse.statusCode === 200) {
            //Retrieve Payout Item details
            let payoutItemId = getResponse.result.items[0].payout_item_id;
            console.log("Retrieving Payout Item details for id - " + payoutItemId);
            let getItemResponse = await getPayoutItem(payoutItemId, true);
            if (getItemResponse.statusCode === 200) {
                let i = 0;
                //Wail till Payout Batch status becomes SUCCESS to cancel an UNCLAIMED payout.
                //This is just for demonstration, defer using this while integration
                //Note: While integrating use Webhooks to get realtime Payouts Batch and Item status updates
                do {
                    let checkPayoutComplete = await getPayout(payoutId);
                    await sleep(2000);
                    if (checkPayoutComplete.result.batch_header.batch_status === "SUCCESS") {
                        //Cancel UNCLAIMED payout item
                        console.log("Cancelling unclaimed payout item for id - " + payoutItemId);
                        let cancelResponse = await cancelPayoutItem(payoutItemId, true);
                        if (getItemResponse.statusCode === 200) {
                            console.log("Unclaimed payout item cancelled successfully for id - " + payoutItemId);

                            //Run cancel failure scenario
                            console.log("Simulate failure on cancelling an already cancelled Payout item with id: " + payoutItemId);
                            await cancelPayoutItem(payoutItemId, true);
                        } else {
                            console.error("Failed to cancel unclaimed payout item for id - " + payoutItemId);
                        }
                        break;
                    }
                    i++;
                } while (i < 5);
                if (i === 5) {
                    console.error("Batch hasn't processed successfully yet!!");
                }
            } else {
                console.error("Failed to retrieve payout item details for id - " + payoutItemId);
            }
        } else {
            console.error("Failed to retrieve payout details for id - " + payoutId);
        }

    } else {
        console.error("Failed to create payout");
    }

    //Execute all failure cases
    console.log("Create a payout with validation failure");
    await createPayoutFailure(true)
    console.log("Retrieving an invalid payout");
    await getPayout("DUMMY", true)
    console.log("Retrieving an invalid payout item");
    await getPayoutItem("DUMMY", true)

    process.exit();

})();

samples/getPayout.js
'use strict';

/**
 * PayPal SDK dependency
 */
const payoutsSdk = require('@paypal/payouts-sdk');

/**
 * PayPal HTTP client dependency
 */
const payPalClient = require('./Common/payPalClient');

/**
 * This dependency is used to create a Payout.
 */
const createPayout = require('./createPayout').createPayout;

/**
 * Retrieves the details for a Payout Batch(GET - /v1/payments/payouts/<batch-id>)
 * This endpoint supports pagination and retrieves only 1000 items in a call. Use pagination links to retrieve all items.
 * Use totalRequired to get the total number of pates
 * @param {string} batchId - the batch id to retrieve
 * @param {boolean} debug - prints debug logs with details on response
 */
async function getPayout(batchId, debug = false) {
    try {
        const request = new payoutsSdk.payouts.PayoutsGetRequest(batchId);
        //Optional, By default pageSize is set to 1000, page is set to 1
        request.page(1);
        request.pageSize(10);
        request.totalRequired(true);

        const response = await payPalClient.client().execute(request);
        if (debug) {
            console.log("Status Code: " + response.statusCode);
            console.log("Payout Batch ID: " + response.result.batch_header.payout_batch_id);
            console.log("Payout Batch Status: " + response.result.batch_header.batch_status);
            console.log("Items count: " + response.result.items.length);
            console.log("First item id: " + response.result.items[0].payout_item_id);
            console.log("Links: ");
            response.result.links.forEach((item, index) => {
                let rel = item.rel;
                let href = item.href;
                let method = item.method;
                let message = `\t${rel}: ${href}\tCall Type: ${method}`;
                console.log(message);
            });
            // To toggle print the whole body comment/uncomment the below line
            console.log(JSON.stringify(response.result, null, 4));
        }
        return response;
    }
    catch (e) {
        if (e.statusCode) {
            if (debug) {
                console.log("Status code: ", e.statusCode);
                const error = JSON.parse(e.message)
                console.log("Failure response: ", error)
                console.log("Headers: ", e.headers)
            }
        } else {
            console.log(e)
        }
    }
}

/**
 * This is the driver function which invokes the getPayout function to retrieve
 * an Payout Batch.
 */
if (require.main === module) {
    (async () => {
        let response = await createPayout();
        await getPayout(response.result.batch_header.payout_batch_id, true);
        await getPayout("DUMMY", true);
    })();
}

/**
 * Exports the Get Payout function. If needed this can be invoked from the
 * order modules to execute the end to flow like create payout, retrieve payout, retrieve payout item
 * and cancel payout item(Optional)
 */
module.exports = { getPayout: getPayout };

samples/getPayoutItem.js
'use strict';

/**
 * PayPal SDK dependency
 */
const payoutsSdk = require('@paypal/payouts-sdk');

/**
 * PayPal HTTP client dependency
 */
const payPalClient = require('./Common/payPalClient');

/**
 * This dependency is used to create a Payout.
 */
const createPayout = require('./createPayout').createPayout;

/**
 * This dependency is used to retrieve a Payout.
 */
const getPayout = require('./getPayout').getPayout;

/**
 * Retrieves a Payout Item details(GET - /v1/payments/payouts-items/<item-id>)
 * @param {string} itemId - the item id to retrieve 
 * @param {boolean} debug - prints debug logs with details on response
 */
async function getPayoutItem(itemId, debug = false) {
    try {
        const request = new payoutsSdk.payouts.PayoutsItemGetRequest(itemId);

        const response = await payPalClient.client().execute(request);
        if (debug) {
            console.log("Status Code: " + response.statusCode);
            console.log("Payout Item ID: " + response.result.payout_item_id);
            console.log("Payout Item Status: " + response.result.transaction_status);
            console.log("Links: ");
            response.result.links.forEach((item, index) => {
                let rel = item.rel;
                let href = item.href;
                let method = item.method;
                let message = `\t${rel}: ${href}\tCall Type: ${method}`;
                console.log(message);
            });
            // To toggle print the whole body comment/uncomment the below line
            console.log(JSON.stringify(response.result, null, 4));
        }
        return response;
    }
    catch (e) {
        if (e.statusCode) {
            if (debug) {
                console.log("Status code: ", e.statusCode);
                const error = JSON.parse(e.message)
                console.log("Failure response: ", error)
                console.log("Headers: ", e.headers)
            }
        } else {
            console.log(e)
        }
    }
}

/**
 * This is the driver function which invokes the getPayoutItem function to retrieve
 * a Payout Item.
 */
if (require.main === module) {
    (async () => {
        let createResponse = await createPayout();
        let getResponse = await getPayout(createResponse.result.batch_header.payout_batch_id);
        await getPayoutItem(getResponse.result.items[0].payout_item_id, true);
        await getPayoutItem("DUMMY", true);
    })();
}

/**
 * Exports the Get Payout Item function. If needed this can be invoked from the
 * order modules to execute the end to flow like create payout, retrieve payout, retrieve payout item
 * and cancel payout item(Optional)
 */
module.exports = { getPayoutItem: getPayoutItem };

samples/cancelPayoutItem.js
'use strict';

/**
 * PayPal SDK dependency
 */
const payoutsSdk = require('@paypal/payouts-sdk');

/**
 * PayPal HTTP client dependency
 */
const payPalClient = require('./Common/payPalClient');

/**
 * Cancels a Payout Item details(POST - /v1/payments/payouts-items/<item-id>/cancel)
 * Only items in status UNCLAIMED can be cancelled to return the funds back to the sender
 * Note: Batch should be in status SUCCESS indicating all payouts items are processed. Use Webhooks to get realtime status updates
 * @param {string} itemId - the item id to retrieve 
 * @param {boolean} debug - prints debug logs with details on response
 */
async function cancelPayoutItem(itemId, debug = false) {
    try {
        const request = new payoutsSdk.payouts.PayoutsItemCancelRequest(itemId);

        const response = await payPalClient.client().execute(request);
        if (debug) {
            console.log("Status Code: " + response.statusCode);
            console.log("Payout Item ID: " + response.result.payout_item_id);
            console.log("Payout Item Status: " + response.result.transaction_status);
            console.log("Links: ");
            response.result.links.forEach((item, index) => {
                let rel = item.rel;
                let href = item.href;
                let method = item.method;
                let message = `\t${rel}: ${href}\tCall Type: ${method}`;
                console.log(message);
            });
            // To toggle print the whole body comment/uncomment the below line
            console.log(JSON.stringify(response.result, null, 4));
        }
        return response;
    }
    catch (e) {
        if (e.statusCode) {
            if (debug) {
                console.log("Status code: ", e.statusCode);
                const error = JSON.parse(e.message)
                console.log("Failure response: ", error)
                console.log("Headers: ", e.headers)
            }
        } else {
            console.log(e)
        }
    }
}

/**
 * Exports the Cancel Payout Item function. If needed this can be invoked from the
 * order modules to execute the end to flow like create payout, retrieve payout, retrieve payout item
 * and cancel payout item(Optional)
 */
module.exports = { cancelPayoutItem: cancelPayoutItem };

samples/createPayout.js
'use strict';

/**
 * PayPal SDK dependency
 */
const payoutsSdk = require('@paypal/payouts-sdk');

/**
 * PayPal HTTP client dependency
 */
const payPalClient = require('./Common/payPalClient');

/**
 * Builds the request body for creating Payouts Batch with 5 Payout Items
 */
function buildRequestBody(includeValidationFailure) {
  let senderBatchId = "Test_sdk_" + Math.random().toString(36).substring(7);
  let amount = includeValidationFailure ? "1.0.0" : "1.00";
  return {
    "sender_batch_header": {
      "recipient_type": "EMAIL",
      "email_message": "SDK payouts test txn",
      "note": "Enjoy your Payout!!",
      "sender_batch_id": senderBatchId,
      "email_subject": "This is a test transaction from SDK"
    },
    "items": [{
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": amount
      },
      "receiver": "payout-sdk-1@paypal.com",
      "sender_item_id": "Test_txn_1"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": amount
      },
      "receiver": "payout-sdk-2@paypal.com",
      "sender_item_id": "Test_txn_2"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": amount
      },
      "receiver": "payout-sdk-3@paypal.com",
      "sender_item_id": "Test_txn_3"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": amount
      },
      "receiver": "payout-sdk-4@paypal.com",
      "sender_item_id": "Test_txn_4"
    }, {
      "note": "Your 1$ Payout!",
      "amount": {
        "currency": "USD",
        "value": amount
      },
      "receiver": "payout-sdk-5@paypal.com",
      "sender_item_id": "Test_txn_5"
    }]
  }
}

/**
 * This function is used to create a Payouts Batch(POST - /v1/payments/payouts)
 * A maximum of 15000 payout items are supported in a single batch request
 * @param {boolean} debug - prints debug logs with details on response 
 */
async function createPayout(debug = false) {
  try {
    const request = new payoutsSdk.payouts.PayoutsPostRequest();
    request.requestBody(buildRequestBody(false));

    const response = await payPalClient.client().execute(request);
    if (debug) {
      console.log("Status Code: " + response.statusCode);
      console.log("Payout Batch ID: " + response.result.batch_header.payout_batch_id);
      console.log("Payout Batch Status: " + response.result.batch_header.batch_status);
      console.log("Links: ");
      response.result.links.forEach((item, index) => {
        let rel = item.rel;
        let href = item.href;
        let method = item.method;
        let message = `\t${rel}: ${href}\tCall Type: ${method}`;
        console.log(message);
      });
      // To toggle print the whole body comment/uncomment the below line
      console.log(JSON.stringify(response.result, null, 4));
    }
    return response;
  }
  catch (e) {
    console.log(e)
  }
}

/**
 * This function is used to create a Payouts Batch(POST - /v1/payments/payouts)
 * with validation failure
 * @param {boolean} debug - prints debug logs with details on response 
 */
async function createPayoutFailure(debug = false) {
  try {
    const request = new payoutsSdk.payouts.PayoutsPostRequest();
    request.requestBody(buildRequestBody(true));

    await payPalClient.client().execute(request);
  }
  catch (e) {
    if (e.statusCode) {
      if (debug) {
        console.log("Status code: ", e.statusCode);
        const error = JSON.parse(e.message)
        console.log("Failure response: ", error)
        console.log("Headers: ", e.headers)
      }
    } else {
      console.log(e)
    }
  }
}

/**
 * This is the driver function which invokes the createPayout function to create
 * a Payout Batch.
 */
if (require.main === module) {
  (async () => {
    await createPayout(true)
    await createPayoutFailure(true)
  })();
}

/**
 * Exports the Create Payout function. If needed this can be invoked from the
 * order modules to execute the end to flow like create payout, retrieve payout, retrieve payout item
 * and cancel payout item(Optional)
 */
module.exports = {
  createPayout: createPayout,
  createPayoutFailure: createPayoutFailure
};